### ISSUE TITLE
The Quality Initiative

### ISSUE THREAD

**Issue by benogle** [atom/atom#7995]
The Atom team will be focusing on quality for the next month or so. This effort is timeboxed: the plan is to finish around the __beginning of Sept__. This means __we probably won't get everything done on this list__ during the allotted time window. After this time window, unfinished things will work into our normal issue handling routine.

### What does this mean?

The goal of the quality initiative is to focus on fixing bugs and adding features that impact a lot of people, but are reasonably low effort. Why? There are a lot of low effort things affecting many people, and a lot of issues that slipped through the cracks while we were focused on the API and 1.0. We're just giving our selves some time to focus on them.

These will be small projects, ideally 1-2 days max. We will avoid rewrites or major refactoring.

The quality initiative will be our main focus, but we will still handle issues and review PRs&mdash;think of this effort as our 80% time.

### What about X bug/feature?

After this quality initiative, we plan on shifting our 80% time to focus on things like [performance](https://github.com/atom/atom/issues/307), better support for [international users](https://github.com/atom/atom-keymap/issues/37), and some features we have in mind.

If you know of a bug or feature that is high-impact / low-effort, not in the list below, and does not fall under performance or internationalization, comment on this issue and we'll figure out if it fits in.

#### Bug

* [ ] Fix largest uncaught errors in bugsnag
* [ ] Continue fixing uncaught errors from https://github.com/atom/atom/issues/7740
  * [x] Cannot read property 'dispose' of undefined https://github.com/atom/tabs/issues/183
  * [x] Uncaught TypeError: Cannot read property 'isEmpty' of undefined https://github.com/atom/bracket-matcher/issues/102
* [x] Fix erasing files when no disk space 
  * [x] https://github.com/atom/atom/issues/3158 @nathansobo
* [ ] Fix app auto-update issues
  * [ ] https://github.com/atom/atom/issues/7833
  * [ ] https://github.com/atom/atom/issues/6612
  * [ ] https://github.com/atom/atom/issues/7854
  * [ ] better feedback https://github.com/atom/atom/issues/8038
* [ ] Fix app update and release notes issues
  * [ ] https://github.com/atom/atom/issues/5692
  * [ ] https://github.com/atom/atom/issues/6922
  * [ ] https://github.com/atom/atom/issues/3757
  * [ ] https://github.com/atom/atom/issues/6482
* [ ] Investigate issues when changing branches
  * [ ] https://github.com/atom/atom/issues/4982
  * [ ] https://github.com/atom/atom/issues/6210
* [x] Cannot open files or folders on Mac OS X El Capitan beta https://github.com/atom/atom/issues/7811 (:heart: :apple:)
  * Likely fixed with upgrade to latest atom shell
* [ ] Text editor interaction problems
  * [ ] Line up/down with multiple selections is broken https://github.com/atom/atom/issues/2470
  * [ ] Beeping when using ctrl-cmd-down
    * @zcbenz? https://github.com/atom/atom/issues/1669
  * [x] ctrl + up/down inserts  or  https://github.com/atom/atom/issues/6949
  * [x] scrolling behavior while mouse-selecting whole lines of text is hard to control https://github.com/atom/atom/issues/6675 @nathansobo (PR #8123)
  * [ ] Selecting text with option + shift + arrow behaves differently to other Mac apps https://github.com/atom/atom/issues/6485
  * [x] Undo of actions taken by package results in corrupted display of buffer https://github.com/atom/atom/issues/7516
  * [x] Undo/redo doesn't change cursor position after multiple block edit https://github.com/atom/atom/issues/8178 - @maxbrunsfeld
  * [x] Clicking and dragging up in gutter does not scroll https://github.com/atom/atom/issues/2696
  * [x] Page Up and Page Down don't move correctly with the window. https://github.com/atom/atom/issues/8320
  * [x] Multi line mouse select breaks in text editor when editor is small enough https://github.com/atom/atom/issues/8539 @nathansobo
  * [x] Strange selection problem https://github.com/atom/atom/issues/8369
  * [ ] Overlay items flicker when near the edge of the window https://github.com/atom/atom/issues/8295
  * [ ] Shrink or scroll editor when panels at bottom open up https://github.com/atom/atom/issues/8068
  * [x] cmd-l in multi-line selection doesn't select beginning of first line  https://github.com/atom/atom/issues/3624
* [x] Keybinding Issues
  * [x] Bundled prefixed keys bindings delay even if 'unset!' by user keymap https://github.com/atom/atom-keymap/issues/78
  * [x] Arpeggio key bindings (all these are related):
    * [x] https://github.com/atom/vim-mode/issues/221
    * [x] https://github.com/atom/vim-mode/issues/334
    * [x] https://github.com/atom/atom-keymap/issues/17
* [ ] Rendering issues
  * [x] Showing line numbers disables sub-pixel AA on tabs, status-bar etc. https://github.com/atom/atom/issues/7904
  * [ ] Thin or small fonts rendered on Windows appear colorfully blurry. https://github.com/atom/atom/issues/7261
* [ ] Window state
  * [ ] Reloading on an external monitor moves the window back to the primary monitor https://github.com/atom/atom/issues/6268
* [x] Open a file at a specific line https://github.com/atom/atom/issues/7342

#### Feature

* [x] Upgrade to latest electron:
  * [x] https://github.com/atom/atom/pull/7877
  * [x] Fixes rendering issues? https://github.com/atom/atom/issues/4924
* [ ] Crash recovery: https://github.com/atom/atom/issues/4161 @nathansobo
* [x] Line endings config option https://github.com/atom/atom/issues/2822 @jlord @maxbrunsfeld
* [x] editor.tabType setting https://github.com/atom/atom/issues/3719 @benogle PR: https://github.com/atom/atom/pull/8356
* [x] Make file type -> language mapping locally configurable https://github.com/atom/atom/issues/1718 @maxbrunsfeld
* [x] Minimize DOM polling to improve idle battery life https://github.com/atom/atom/issues/4019 @nathansobo @jlord
* [ ] Remove deprecated code https://github.com/atom/atom/issues/5164
* [ ] Do something about metrics visibiliity https://github.com/atom/atom/issues/4966
* [x] Async services https://github.com/atom/service-hub/issues/7
* [ ] Improve auto-indentation
  * [ ] https://github.com/atom/atom/issues/5473
  * [ ] https://github.com/atom/atom/issues/6655
  * [ ] https://github.com/atom/atom/labels/auto-indent
* [ ] Autosave untitled buffers https://github.com/atom/atom/issues/942
* [ ] Ctrl-tab to be MRU: https://github.com/atom/tabs/issues/17
* [ ] "Save" or "Save All" on unmodified, opened files results in "Unable to save file: permission denied" error
  * [ ] https://github.com/atom/atom/issues/7566
  * [ ] https://github.com/atom/atom/issues/2871
* [ ] Make preview-tabs really good https://github.com/atom/tabs/issues/170

### Atom.io

* [ ] Docs for 1.0.5 and 1.0.6 are not shown https://github.com/atom/docs/issues/111
* [ ] Make it low or no maintenance
* [ ] atom:// url package installation
  * [ ] https://github.com/atom/atom/issues/5262
  * [x] https://github.com/atom/settings-view/issues/436
* [ ] flight manual apparently has a lot of small bugs:
  * [ ] https://github.com/atom/docs/issues/94
  * [ ] https://github.com/atom/docs/issues/85
  * [ ] https://github.com/atom/docs/issues/81
  * [ ] https://github.com/atom/docs/issues/74

#### Find and Replace

* [x] handling path negation https://github.com/atom/find-and-replace/issues/149
  * PR https://github.com/atom/atom/pull/8071
* [x] Project replace undo (just add an alert)
* [x] Performance in large files (just dont auto-add markers when over some large amt of results) https://github.com/atom/find-and-replace/issues/174
  * Added a min length setting before live searching https://github.com/atom/find-and-replace/pull/469
  * There arent really any other quick wins. After this initiative, we will just make it faster.
* [x] Make cmd-d more consistent. Everyone thinks this is a bug https://github.com/atom/find-and-replace/pull/482
  * [x] Maybe this: https://github.com/atom/atom/issues/6800#issuecomment-121976136
  * [x] https://github.com/atom/find-and-replace/issues/340
  * [x] https://github.com/atom/find-and-replace/pull/430
  * [x] https://github.com/atom/find-and-replace/pull/356
  * [x] https://github.com/atom/find-and-replace/pull/450
* [x] Find settings are not shared between project find and buffer find https://github.com/atom/find-and-replace/issues/431
  * https://github.com/atom/find-and-replace/pull/475
* [x] Finish muan's sort PR https://github.com/atom/find-and-replace/pull/320

#### Settings View

* [ ] Better loadPackages system; keep cache of packages on system, all views use this cache https://github.com/atom/settings-view/issues/633
* [ ] Better loadOutdated system; keep cache of outdated packages https://github.com/atom/settings-view/issues/632
  * [ ] Fix: https://github.com/atom/settings-view/issues/409
* [ ] Updating a disabled package automatically re-enables that package https://github.com/atom/settings-view/issues/603
* [ ] Bug fixing pass
* [ ] Render setting description links as links https://github.com/atom/settings-view/issues/618

#### Autocomplete

* [x] CSS
  * [x] Pops up in non-ideal situations
    * [x] https://github.com/atom/autocomplete-css/issues/15
    * [x] https://github.com/atom/autocomplete-css/issues/18
    * [x] https://github.com/atom/autocomplete-css/issues/20
    * [x] https://github.com/atom/autocomplete-css/issues/21
    * [x] https://github.com/atom/autocomplete-plus/issues/513
  * [x] SASS support https://github.com/atom/autocomplete-css/pull/13
* [x] Actually for real fix html escaping: https://github.com/atom/autocomplete-plus/issues/536
* [x] Figure out something nice for the core keymap issues https://github.com/atom/autocomplete-plus/pull/543
  * https://github.com/atom/autocomplete-plus/issues/319
* [x] Page up/down
  * https://github.com/atom/autocomplete-plus/issues/503

#### Fuzzy Finder

* [ ] Take a pass at indexing, maybe use scandal
  * https://github.com/atom/fuzzy-finder/issues/112
  * https://github.com/atom/fuzzy-finder/issues/88
  * https://github.com/atom/fuzzy-finder/issues/53
  * https://github.com/atom/fuzzy-finder/issues/30
* [ ] 100% CPU usage when started in the home folder https://github.com/atom/atom/issues/6437

#### Bracket Matcher

* [ ] Surrounding multiple words with \` characters is broken https://github.com/atom/atom/issues/6940

### Modules

* [ ] Improve fuzzaldrin
  * https://github.com/atom/fuzzaldrin/issues/20
  * https://github.com/atom/fuzzaldrin/pull/22

### Pull requests to review

* [x] Babel safeguard https://github.com/atom/atom/pull/8182
* [x] Makefile grammar improvement https://github.com/atom/language-make/pull/17
* [x] Referencing patterns in other grammars https://github.com/atom/first-mate/pull/53
* [x] Resizing tree view with other panels https://github.com/atom/tree-view/pull/534
* [x] Copy selection command https://github.com/atom/atom/pull/8025
* [ ] Infinite loop in parser https://github.com/atom/first-mate/pull/47
* [x] Resize panes with keyboard https://github.com/atom/atom/pull/7354
* [x] Clean up classes when switching themes https://github.com/atom/one-dark-ui/pull/87
* [ ] #{} interpolation improvement https://github.com/atom/bracket-matcher/pull/161
* [ ] Enum description in config https://github.com/atom/atom/pull/7916 and https://github.com/atom/settings-view/pull/589
* [x] Collapsing sections in settings view: https://github.com/atom/settings-view/pull/593

**Comment by Mericke** [comment]
Upgrading to LESS 2.5.1 would be nice as Atom is almost a year behind the current version at this point. I personally need it to fix a few things as do others i'm sure.

There already seems to be an active pull request that fixes the 2.x changes here: https://github.com/atom/less-cache/pull/8.

I've tested it on a locally compiled Atom using Windows 10 and haven't run into any issues yet so i'm pretty this at least qualifies for low effort.

**Comment by hotoiledgoblinsack** [comment]
A minor feature that I've been sorely missing for some time is https://github.com/atom/find-and-replace/issues/369 to be able to go back when I've accidentally pressed the return key too many times when searching for occurrences.

**Comment by 190n** [comment]
What about atom/bracket-matcher#80?

**Comment by mehcode** [comment]
Add a method to the API to switch the active project: https://github.com/atom/atom/issues/6984

----

Would allow for a sublime-style project switcher that I'm sure many of us were used to in sublime.

**Comment by ObviouslyGreen** [comment]
Dragging and dropping a file from the file system occasionally opens it as Chromium instead of Atom #4300 

Having to restart atom after quickly dragging and dropping a file just to check something can be annoying at times, not sure if it affects everyone though since the issue report isn't super active.

**Comment by adambuczynski** [comment]
This issue is possibly related to the fuzzy finder indexing issues: https://github.com/atom/atom/issues/7513

**Comment by Stargator** [comment]
Maybe something concerning right clicking on a folder to create a file should automatically have the directory prepended without displaying as such to the user via the popup. As a new user when I saw that, I thought it was "wrong" and replaced it with just the name of the file as that has been the usually UI protocol for most, if not all, of other applications.

**Comment by adambuczynski** [comment]
I actually like that feature. It allows me to change my mind if I decide to
create the file elsewhere, and makes it easy to copy paste the full path is
needed.

**Comment by gyf304** [comment]
I'd really appreciate fixing the 30 second freeze on windows platform. #7275

**Comment by Stargator** [comment]
@adambuczynski  Hmm, then I guess someone could look into making a plugin that toggles it on/off, if it isn't implemented in the main branch.

I always find once I decide where to put a file, I'm happy with it for a long while, barring any need to re-organize the project.

**Comment by gut4** [comment]
What about "Non-native" Fullscreen on OSX? https://github.com/atom/atom/issues/967

**Comment by Stanzilla** [comment]
Upgrading Squirrel should be part of `Fix app update and release notes issues`, no?

**Comment by teresko** [comment]
The #2822 has been the **only reason** why I have not been able to migrate to Atom. It's nice to see, that it might change soon.

**Comment by Victorystick** [comment]
I nominate the auto-indentation issue #6655. It seems suitable for the quality initiative as pretty much all language grammars seem affected.

**Comment by faddah** [comment]
as an update, the issue in #7833 & #6612 about updating is _still_ happening. this is on a 2009 iMac 27", intel core i7 2.8 GHz processor, 16 GB RAM, Mac OS X 10.10.4 Yosemite, and Atom Editor update going from v.1.0.2 to v.1.0.3.  still happending, says it's downloaded the update and needs to restart, hit "restart," nothing happens. launch Atom Editor again, starts same process all over again, never updates. thought you'd want to know.

best,

—  faddah
     portland, oregon, u.s.a.

**Comment by tiemay** [comment]
This cursor issue is pretty annoying, I don't know how much work would be involved to fix it though.

#3130, #3278, #3622

**Comment by jugglingnutcase** [comment]
i'd like to advocate for https://github.com/atom/find-and-replace/issues/309, which i started, that i think may be related to the find and replace issues with searching when there's a dot in the name of the folder.

**Comment by blah238** [comment]
Please add https://github.com/atom/atom/issues/6884 (Middle mouse button scrolling in the text editor) to the list, this is a pretty major usability issue for me.

**Comment by bronson** [comment]
Wondering about https://github.com/atom/atom/issues/2203 (git integration only works if .git is in the root dir).  Worth including in this bugsmash?

**Comment by 1j01** [comment]
[Moving lines](https://github.com/atom/atom/issues/7653) and [context menu context](https://github.com/atom/atom/issues/3502); also [outdenting](https://github.com/atom/atom/issues/6362)

**Comment by ThibaudDauce** [comment]
@tiemay +1 for cursor issues : #3130, #3278, #3622

**Comment by xeron** [comment]
What about grammar detection?

I have serious troubles using Atom with both Ruby and Rails projects because `.rb` files sometimes being opened with `Ruby` and sometimes with `Ruby on Rails` syntax.

Related report: https://github.com/atom/language-ruby-on-rails/issues/34

**Comment by solarsailer** [comment]
I don't know the time it might take, but a recently-used/edited ordered fuzzy finder would be immensely useful.

I navigate almost solely through the fuzzy finder, and each time it shows me the exact same list in the same order even if I have never edited the first file of the list.

Sublime Text handles this pretty nicely.

The number of edits and open should be enough to rank a file higher than the others.

**Comment by bronson** [comment]
@solarsailer if `cmd-b` doesn't do what you want, you should open a new issue (search for existing ones of course!).  This is for tracking other issues.

**Comment by solarsailer** [comment]
`cmd+b` is only for already opened file. ;)

![example](https://cloud.githubusercontent.com/assets/478621/9270429/11193f64-4272-11e5-808e-f4c2810f3624.png)

If I close my buffers, or if I quit/restart Atom, or if I come back later to edit the `_entry.scss` file, I have to go through all the results.

Atom should know that I edit `_entry.scss` quite often, then `_entry.erb`, then `entry.rb`. The fuzzy finder should evolve and take my habit into account.

**Comment by jensljungblad** [comment]
Agreed about the grammar detection: https://github.com/atom/language-ruby-on-rails/issues/34

**Comment by Ryex** [comment]
I just want to point out that the item about line endings referring to #2822 has a check mark marking it as complete. it's not and should be unmarked so that it doesn't get overlooked.

thanks for this great work.

**Comment by maxbrunsfeld** [comment]
@Ryex sorry for the confusion: we check these once we ship the code to master. The line endings feature is on Atom master, it just hasn't been released yet. Atom 1.0.8 is coming soon though!

**Comment by billyjanitsch** [comment]
> If you know of a bug or feature that is high-impact / low-effort, not in the list below, and does not fall under performance or internationalization, comment on this issue and we'll figure out if it fits in.

Would you consider taking a look at #3624? It's a frustrating, old bug that doesn't seem like it would be hard to fix.

**Comment by nathansobo** [comment]
@billyjanitsch Thanks for the heads up. #3624 will be fixed on the next release.

**Comment by m1ga** [comment]
An option "use restore session / ignore old session" would be also something to have a look at: https://github.com/atom/atom/issues/6605 20 participants) since not all users want to start with an old session.
I don't know about the exact code parts but an if-function around the 'load old session' part shouldn't be a big feature (I hope so :grin: )

**Comment by solbjorn** [comment]
In addition to @m1ga post, it would be very nice to see an option to restore saved session EVEN if launched WITH file arguments (like Sublime does), see #7186.

**Comment by adambuczynski** [comment]
This issue for broken icon images seems like a good candidate for the quality initiative:

https://github.com/atom/settings-view/issues/327

It's rather ugly with the broken images, and deleting the cached files doesn't help:

![ss](https://www.dropbox.com/s/zqnix04lo2v5snr/Screenshot%202015-09-09%2007.29.52.png?dl=1)

**Comment by benogle** [comment]
Closing this out. Most of the team's efforts are now dedicated toward performance. The rest of the unfinished items will be worked into our normal bug-fixing / enhancement-adding process.

**Comment by nathansobo** [comment]
We'll do another focused pass like this once some more performance improvements land.

**Comment by adambuczynski** [comment]
That's a pity, there's a lot of issues lingering in here still that would benefit from one last push :)

**Comment by benogle** [comment]
> That's a pity, there's a lot of issues lingering in here still that would benefit from one last push :)

Agreed. But we only have so many people!

**Comment by adambuczynski** [comment]
Clearly need more :D

**Comment by benogle** [comment]
> Clearly need more :D

Working on it...

**Comment by JesseLeite** [comment]
Thank you all for your work on Atom! 💖

**Comment by jeancroy** [comment]
@benogle 

I'd like to propose grandfathering https://github.com/atom/fuzzaldrin/pull/22 from The Quality Initiative to the Performance one.

After a few performance push it's now 2x faster than classic fuzzaldrin on official benchmark query. It's also 2X faster in a variety of condition and I'm tempted to say on average.

In addition to being faster it could close about 6 open issues including 2 that act as catch-all. It play a central role in autocomplete, fuzzy-finder and command palette,  thus in the overall atom experience.

To help review process I've included a documentation for the PR, including a section on benchmark and addressed issue. (near the end)

It's a large PR, by this time a rewrite, it may take a few day to review it. However it is more accurate, faster, backward compatible, and by now there's probably the equivalent of a month of full time work invested in it.

 Even outside an official review process it would be great to have some feedback. The impression of shooting in the dark is a bit discouraging.
